bsmetadata/preprocessing_utils.py

```diff
@@ -716,6 +716,103 @@ def preprocess(self, examples: Dict[str, List]) -> Dict[str, List[OneToOneFeatur
         return examples
 
 
+class EntityParagraphPreprocessor(MetadataPreprocessor):
+    """A metadata preprocessor for updating entity information based on paragraphs."""
+
+    def __init__(
+        self, col_to_store_metadata="metadata", col_entity="metadata_entity", col_paragraph="metadata_paragraph"
+    ) -> None:
+        self.col_entity = col_entity
+        self.col_paragraph = col_paragraph
+        super().__init__(col_to_store_metadata=col_to_store_metadata)
+
+    @property
+    def new_columns_minimal_features(self) -> Dict[str, List[OneToOneFeature]]:
+        features = {
+            self.col_to_store_metadata: [
+                {
+                    "char_end_idx": Value("int64"),
+                    "char_start_idx": Value("int64"),
+                    "key": Value("string"),
+                    "relative_end_pos": Value("int64"),
+                    "relative_start_pos": Value("int64"),
+                    "type": Value("string"),
+                    "value": Value("string"),
+                }
+            ],
+        }
+        return features
+
+    def preprocess(self, examples: Dict[str, List]) -> Dict[str, List[OneToOneFeature]]:
+        example_metadata_list = (
+            examples[self.col_to_store_metadata]
+            if self.col_to_store_metadata in examples
+            else [[] for _ in range(len(examples[self.col_entity]))]
+        )
+
+        # Iterate through the metadata associated with all examples in this batch.
+        for example_entity, example_paragraph, example_metadata in zip(
+            examples[self.col_entity], examples[self.col_paragraph], example_metadata_list
+        ):
+
+            if not example_entity or not example_paragraph:
+                continue
+
+            # Iterate through the entities associated with this example.
+            for entity in example_entity:
+                # If entity["key"] != "entity", we skip this entity. Added this check since tests were failing.
+                if entity["key"] != "entity":
+                    continue
+                # Initialize the start and end index of an entity
+                start_index = entity["char_start_idx"]
+                end_index = entity["char_end_idx"]
+                check = True
+
+                # Search the start and end index of paragraph in between which the entity is present without any duplicate entity values.
+                for paragraph in example_paragraph:
+                    # If paragraph["key"] != "paragraph", we skip this paragraph. Added this check since tests were failing.
+                    if paragraph["key"] != "paragraph":
+                        check is False
```

Comment:
```suggestion
                        check = False
```
I guess this is the indented assignment?

Reply:
Code quality test was failing because of this.

Reply:
@manandey For https://github.com/bigscience-workshop/metadata/runs/6047523378?check_suite_focus=true#step:5:13 , I guess it was `check == False` or `if check == False` and yet it seemed an assignment rather than a comparison. When `check is False` without an `if`, `check` isn't changed.